### PR TITLE
Improve Megatron Tokenization: streaming, reasoning_content support, HF in-memory tokenization, etc

### PR DESCRIPTION
**Pull request by kevalmorabia97**
### What does this PR do?

Type of change: New feature

Improvements to `megatron_preprocess_data` for Nemotron v3 post-training datasets and Megatron-Bridge distillation workflows:

- **`--reasoning_content`** flag (`strip` / `inline` / `native`) to handle the `reasoning_content` field in Nemotron Post-Training v3 assistant messages
- **No intermediate JSONL** for HuggingFace datasets — load directly from Arrow cache via `_iter_hf_as_json` + `process_hf_split`
- **Return output prefixes** (`list[str]`) from the Python API so callers can build `--data_paths` without hardcoding paths; also printed at end of run
- **Gzip input support** — `.jsonl.gz` files accepted directly; `--input_dir` globs both `*.jsonl` and `*.jsonl.gz`
- **`--strip_newlines`** flag (opt-in) to replace newlines with spaces in plain-text values; default preserves newlines (no breaking change for code/structured-text datasets)
- **`--hf_streaming`** flag for very large datasets — only consumed rows are downloaded; automatically falls back to non-streaming (with a warning) if `--hf_max_samples_per_split` is not set, since streaming without a cap is slower than cached non-streaming
- **Auto-shuffle** when `--hf_max_samples_per_split` is set — reservoir sampling (buffer=10,000, seed=42) applied before capping to avoid biased prefix sampling
- Remove `_document` suffix from output filenames (`_text.bin` instead of `_text_document.bin`)
- Fix duplicate BOS token for chat-template data (`add_special_tokens=False`)
- Fix `TypeError` in `process_hf_split` (`sum(list)` not `sum(int)`)
- Suppress duplicate prints across pool workers via `_is_main_or_first_worker()`
- Raise `KeyError` instead of warning for missing JSON keys
- Default `hf_split=None` (all splits) instead of `"train"`

### Usage

```python
from modelopt.torch.utils.plugins.megatron_preprocess_data import megatron_preprocess_data

# Nemotron v3 with reasoning content preserved inline as <think>...</think>
prefixes = megatron_preprocess_data(
    hf_dataset="nvidia/Nemotron-Post-Training-Dataset-v3",
    json_keys=["messages"],
    tokenizer_name_or_path="Qwen/Qwen3-0.6B",
    output_dir="tokenized/",
    workers=32,
    reasoning_content="inline",
)
# prefixes == ["tokenized/nvidia--Nemotron-Post-Training-Dataset-v3_..._messages"]
data_paths = [x for p in prefixes for x in ("1.0", p)]

# Large pretraining dataset — stream + cap (auto-shuffled before capping)
prefixes = megatron_preprocess_data(
    hf_dataset="nvidia/Nemotron-CC-v2.1",
    hf_name="High-Quality",
    hf_max_samples_per_split=5_000_000,
    hf_streaming=True,
    json_keys=["text"],
    tokenizer_name_or_path="Qwen/Qwen3-0.6B",
    output_dir="tokenized/",
    workers=32,
    append_eod=True,
    strip_newlines=True,
)
```

### Testing

- New unit tests
- Tested tokenization on Nemotron Pretraining and Post-training v3 datasets

### Before your PR is "*Ready for review*"

Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md) and your commits are signed (`git commit -s -S`).

Make sure you read and follow the [Security Best Practices](https://github.com/NVIDIA/Model-Optimizer/blob/main/SECURITY.md#security-coding-practices-for-contributors) (e.g. avoiding hardcoded `trust_remote_code=True`, `torch.load(..., weights_only=False)`, `pickle`, etc.).

- Is this change backward compatible?: ✅ (output filename changed: `_text_document` → `_text`; existing callers need to re-tokenize or rename files)
- If you copied code from any other sources or added a new PIP dependency, did you follow guidance in `CONTRIBUTING.md`: N/A
- Did you write any new necessary tests?: ✅
- Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?: ✅

## Summary by CodeRabbit

* **New Features**
  * Preprocessing tool: configurable reasoning-content modes (strip|inline|native), optional newline stripping, gzip (.jsonl.gz) input support, HF streaming mode, auto-shuffle when per-split max samples set, returns output-file prefixes, and processes all HF splits by default without writing intermediate JSONL.

* **Documentation**
  * Consolidated and reformatted dataset preparation and tokenization guidance; updated install/auth instructions and examples.

* **Tests**
  * Tests updated to validate returned prefixes, reasoning-content behaviors, gzip input handling, HF streaming warnings, and HF output assertions.

### CODE REVIEW

**Review by ChenhanYu — APPROVED** [review]
# PR Review: Improve Megatron Tokenization

Well-structured feature PR. Streaming mode, reasoning_content handling, direct HF tokenization, gzip support, and return prefixes are all clean additions. Documentation and test coverage are solid.

## Comments

### 1. Output filename format changed — silent breaking change

Old: `{stem}_{key}_document.{bin,idx}` → New: `{stem}_{key}.{bin,idx}`

The `_document` suffix is dropped. Any downstream scripts or configs that hardcode the old naming pattern will break. Consider noting this in the CHANGELOG entry.

### 2. `process_hf_split` pool not closed on exception (`megatron_preprocess_data.py:~413`)

If an exception occurs between `Pool()` and `pool.close()`, the pool leaks. Consider `try/finally` or a context manager. Non-blocking.

### 3. Streaming vs non-streaming shuffle semantics differ

Non-streaming: `split[:{max_samples}]` takes first N rows then shuffles them. Streaming: shuffles the stream then `take(N)`. Both randomize, but may select different subsets. Fine in practice, worth a brief comment if reproducibility matters.

### 4. `final_enc_len` computation differs between file and HF paths

`process_json_file`: `sum(sentence_lens[key])` (one key). `process_hf_split`: `sum(sum(sentence_lens[key]) for key in sentence_lens)` (all keys). For multi-key scenarios, the file path undercounts. Minor.

### 5. `_is_main_or_first_worker` uses private `_identity`

Pragmatic choice for gating log noise but `_identity` is a CPython implementation detail. Worth a brief comment.

LGTM — the filename format change is the only item worth documenting before merge.

This is an AI-assisted review — human sign-off required before merging.
